### DISCUSSION TITLE
Extracting inner products

### DISCUSSION THREAD

**Discussion by braxtonowens** [ACEsuit/ACE1.jl#37] [General]
IPFitting documentation states that: 
"lsqdb_kron.h5 : this stores all the inner products <basis, data> as a single large matrix." 

I am wanting to access these inner products but unfortunately, it is expressed as a 0xN matrix. I did a little bit of skimming online but I am unsure exactly how to access the data stored in a 0XN matrix.

![image](https://user-images.githubusercontent.com/45109437/177190817-1609b339-5e73-4e16-b0a2-15e39e704b28.png)

Background information: I want the ACE expression to see how it compares to other descriptors for representing grain boundaries. 

Happy 4th of July! 

Braxton Owens

**Comment by cortner** [comment]
I'm in holiday but will try to look into it when Im back. I may need more info - 0xN suggests that you assembled over an empty set of observations.

**Comment by braxtonowens** [comment]
Yeah, No rush. Ill look into it. Thanks.

**Comment by braxtonowens** [comment]
It does look like there are 0 observations.
`length(data[1].D)` returns 0 

I am generating my .xyz files by reading LAMMPS dump files then writing as an xyz. I'm sure there is a better way, I might be losing information. Attached is one of the files generated(changed to a .txt for github to accept). 

Any Ideas why no observations are made?

[my_xyz.txt](https://github.com/ACEsuit/ACE1.jl/files/9115108/my_xyz.txt)

![image](https://user-images.githubusercontent.com/45109437/179076277-61dc7813-59dc-42b2-8dfa-5e3957cc4699.png)

**Reply by cortner** [reply]
My best guess is you need to specify the correct keys to read energies, forces, virials. I'm on holiday without laptop so hard for me to look more closely.

@casv2 - would you mind taking a quick look and see if you can help?

**Reply by casv2** [reply]
Yes so usage is as follows:

```al = IPFitting.Data.read_xyz("./fname.xyz", energy_key="energy", force_key="forces", virial_key="virial")```

where the different keys refer to the keys in the .info storing the energy/virial, and .arrays stores the forces. The output says it found several keys in your info and arrays but no energy/forces/virials. Are you sure you've included data in your training database? I think if you examine your .xyz file there is no data to fit, is that right? 

Once the ```missing``` returns 0's you've made sure that the data has successfully been read in by IPFitting

**Reply by braxtonowens** [reply]
My goal is not to train a potential. I just want the feature matrix of my data represented using the ACE basis. I believe this only requires positional and atom type data. Is this possible?

**Reply by casv2** [reply]
Yes apologies I misunderstood, the code below should work

```
using ACE1, IPFitting

al = IPFitting.Data.read_xyz("./fname.xyz", energy_key="", force_key="", virial_key="")

basis = rpi_basis(species = [:C],
      N = 3,       # correlation order = body-order - 1
      maxdeg = 12,  # polynomial degree
      r0 = 1.4,     # estimate for NN distance
      rin = 0.1,
      rcut = 5.5,   # domain for radial basis (cf documentation)
      pin = 2)   

basis_eval1 = site_energy(basis, al[1].at, 2) 
```

The code above evaluates the basis on the second indexed atom in the first configuration. If you want to evaluate the sum over the sites (say to fit the total energy) use the following:

```
basis_eval2 = energy(basis, al[1].at)
```

**Reply by cortner** [reply]
Yes, that's absolutely possible, but IPFitting is probably not set up for that. So you may need to write a short script by hand since OR supply artificial data to "hack" IPFitting. 

First, can you please file an issue with [ACEfit](https://github.com/ACEsuit/ACEfit.jl) which is the successor of IPFitting to explain the functionality you want and how you would like it to be made available. 

Secondly, it might be quickest for you to just do the following: 
```Julia 
basis = .... # generate an ACE basis 
structures = .... # generate or load a few Atoms objects.
A = zeros(length(structures), length(basis))
for (i, at) in enumerate(structures)
   A[i, :] = energy(basis, at)
end 
```
If you have A LOT of structures, then you could multi-thread it (take a quick look at Julia docs, it should be as simple as annotating the loop with `@threads`. 

One comment : this scripts evaluates the total-potential-energy like descriptors. It does NOT give you the per-site descriptors. If that's what you want then that's also easy and I can quickly draft a script for you. 

Warning: Unfortunately my laptop is broken so I can't test anything I'm suggesting here!!!

**Reply by braxtonowens** [reply]
Thank you. I will file an issue. I do actually have many structures, thanks for the advise. I also am looking for a per-atom description of the structure. A script would be greatly appreciated.
